Docs/how to add a new sql dialect (#1497)
great-expectations/great_expectations: docs/how_to_guides/miscellaneous/how_to_add_and_test_a_new_sqlalchemydataset_class.rst

.. _how_to_guides__miscellaneous__how_to_add_and_test_a_new_sqlalchemydataset_class:

How to add support for a new SQLAlchemy dialect
===============================================

This guide will help you extend the Great Expectations execution layer to work on a new SQL dialect.

.. warning::

   This guide is a working checklist. We improve it each time we add support for a new SQL dialect, but it still has rough edges. And adding support for a new database will probably never be a cookie-cutter operation.

   If you're interested in extending Great Expectations in this way, please reach out on `Slack <greatexpectations.io/slack>`__ in the ``#contributors-contributing`` channel, and we'll work with you to get there.

Steps
-----

1. Install and document dependencies
####################################

>     * **Install the python dependencies for your dialect.**  While you're at it, please add a line(s) to ``requirements-dev.txt``. Examples:

Reviewer: ```suggestion
    * **Install the python dependencies for your dialect.**  While you're at it, please add a line(s) to ``requirements-dev-sqlalchemy.txt``. Examples:
```

    
        .. code-block:: bash
        
            pyhive[presto]>=0.6.2
            PyAthena[SQLAlchemy]>=1.1
            pybigquery>=0.4.15

>     * **Install any necessary drivers**. sqlalchemy requires special drivers for some databases. If yours is one of them, please carefully document the steps in {{document TBD}}

Reviewer: ???

Author: arg. Missed this one.

Author: ```suggestion
    * **Install any necessary drivers**. sqlalchemy requires special drivers for some databases. If yours is one of them, please carefully document the steps .
```

The core team will not be able to merge your contribution until they're able to replicate your setup.

2. Make changes to sqlalchemy_dataset.py
########################################

    * **At the top the file, add a try-except block**

        .. code-block:: python

            try:
                import pyhive.sqlalchemy_presto as presto
            except ImportError:
                presto = None

    * **In SqlAlchemyDataset.__init__, add an elif clause**

        .. code-block:: python

            elif self.engine.dialect.name.lower() == "presto":
                self.dialect = import_module("pyhive.sqlalchemy_presto")

    * **In SqlAlchemyDataset.create_temporary_table, add an elif clause**

        .. code-block:: python

            elif (self.engine.dialect.name.lower() == "awsathena"
              or self.engine.dialect.name.lower() == "presto"):
                stmt = "CREATE OR REPLACE TABLE {table_name} AS {custom_sql}".format(
                    table_name=table_name, custom_sql=custom_sql)

> 

Reviewer: Add implementations for your dialect for all relevant _get_dialect_XXX methods

Author: I'm not sure what you're saying here.

Reviewer: The contributor must review and usually add implementations for any method that starts with `_get_dialect` for their dialect.

                try:
                    engine = sa.create_engine(presto_conn_str)
                    conn = engine.connect()
                except (ImportError, sa.exc.SQLAlchemyError):
                    raise ImportError(
                        "presto tests are requested, but unable to connect to the presto database at "
                        f"'{presto_conn_str}'"
                    )
                test_backends += ["presto"]

    * **In the sa fixture method, add your test_backend to the list of backends**

        .. code-block:: python

            if "postgresql" not in test_backends and "sqlite" not in test_backends and "presto" not in test_backends:

    * **For each of the test datasets, add a schema entry for the dialect.**

        Examples:

            * In ``numeric_high_card_dataset`` : ``"presto": {"norm_0_1": "DOUBLE"},``
            * In ``datetime_dataset`` : ``"presto": {"datetime": "TIMESTAMP"},``
            * In ``dataset_sample_data`` : ``"presto": {"infinities": "DOUBLE", "nulls": "DOUBLE", "naturals": "DOUBLE"},``

4. Make changes to tests/test_utils.py
######################################

>     * **Add a try-except clause to import dialect-specific types and map them to generic types.**

Reviewer: Those are *not* supposed to be generic types; they should correspond to the dialect-specific correct type names that will be used in test schema definitions.

Author: Ah, that makes more sense.

Author: ```suggestion
    * **Add a try-except clause to import dialect-specific types and map them to type names that will be used in test schema definitions.**
```

        .. code-block:: python

            try:
                import sqlalchemy.types as sqltypes
                from pyhive.sqlalchemy_presto import presto as prestotypes
                from pyhive.sqlalchemy_presto import PrestoDialect as prestodialect

                PRESTO_TYPES = {
                    "VARCHAR": sqltypes.VARCHAR,
                    "TEXT": sqltypes.VARCHAR,
                    "CHAR": sqltypes.CHAR,
                    "DOUBLE": prestotypes.DOUBLE,
                    "INTEGER": sqltypes.INTEGER,
                    "SMALLINT": sqltypes.SMALLINT,
                    "BIGINT": sqltypes.BIGINT,
                    "DATETIME": sqltypes.TIMESTAMP,
                    "TIMESTAMP": sqltypes.TIMESTAMP,
                    "DATE": sqltypes.DATE,
                    "FLOAT": prestotypes.DOUBLE,
                    "BOOLEAN": prestotypes.BOOLEAN,
                }
            except ImportError:
                PRESTO_TYPES = {}

>     * **In get_dataset, add an elif clause to connect to data, .** Note: some SQL backends require a schema. If so, this is also the right place to create a test schema, usually called ``test_ci``.

Reviewer: This is not about connecting to data, but rather about providing instructions for how to build temporary datasets in your database of choice.

Author: ```suggestion
    * **In get_dataset, add an elif clause to build temporary datasets for testing.** Note: some SQL backends require a schema. If so, this is also the right place to create a test schema, usually called ``test_ci``.
```

                            df[col] = pd.to_numeric(df[col])
                        elif type_ in ["float", "double"]:
                            df[col] = pd.to_numeric(df[col])
                        elif type_ in ["timestamp", "datetime"]:
                            df[col] = pd.to_datetime(df[col]).dt.strftime('%Y-%m-%d %H:%M:%S')
                        elif type_ in ["varchar"]:
                            df[col] = df[col].astype(str)

                tablename = generate_test_table_name(dataset_id)

                conn.execute("CREATE SCHEMA IF NOT EXISTS test_ci")
                df.to_sql(name=tablename, con=conn, index=False)

                # Build a SqlAlchemyDataset using that database
                return SqlAlchemyDataset(
                    tablename, engine=conn, profiler=profiler, caching=caching
                )

    * **Add your dialect to candidate_test_is_on_temporary_notimplemented_list**.

        .. code-block:: python

            def candidate_test_is_on_temporary_notimplemented_list(context, expectation_type):
                if context in ["sqlite", "postgresql", "mysql", "presto"]:

> 5. Get tests to pass

Reviewer: New section: 
Make changes to test definitions 
-> describe adding the dialect to only_for or "suppress_test_for"

Author: IMO, we should discourage contributors from using `only_for` and `suppress_test_for`, since they create exceptions to the "works the same on all execution engines" promise we're trying to adhere to.

Proposal: under "Get tests to pass", add this blurb:

In some rare cases, you may need to suppress certain tests for your SQL backend. In that case, you can use the `only_for` or `suppress_test_for` flags in the test configs. However, we try very hard to avoid such cases, since they weaken the "works the same on all execution engines" principle of Great Expectations.

Reviewer: I like that language for only_for and suppress_test_for! The contributor should also be adding any tests that are needed to highlight dialect-specific quirks.

Instead of "Get tests to pass" how about "Use tests to verify [semantic] consistency".  I *really* hate it when people think of tests as a guarantee that a thing is right, and "get tests to pass" makes it sound like if the tests pass then we think it's right. Tests passing is necessary but not sufficient.

How to add support for a new SQLAlchemy dialect
===============================================

This guide will help you extend the Great Expectations execution layer to work on a new SQL dialect.

.. warning::

   This guide is a working checklist. We improve it each time we add support for a new SQL dialect, but it still has rough edges. And adding support for a new database will probably never be a cookie-cutter operation.

   If you're interested in extending Great Expectations in this way, please reach out on `Slack <greatexpectations.io/slack>`__ in the ``#contributors-contributing`` channel, and we'll work with you to get there.

Steps
-----

1. Install and document dependencies
####################################

    * **Install the python dependencies for your dialect.**  While you're at it, please add a line(s) to ``requirements-dev.txt``. Examples:
    
        .. code-block:: bash
        
            pyhive[presto]>=0.6.2
            PyAthena[SQLAlchemy]>=1.1
            pybigquery>=0.4.15

    * **Install any necessary drivers**. sqlalchemy requires special drivers for some databases. If yours is one of them, please carefully document the steps in {{document TBD}}

>     * **If possible, add docker containers for test resources.** Otherwise, document the steps to set up a suitable test environment. In either case, please document steps in {{document TBD}}.

Reviewer 2: a little pointer to the CI definition here ?

Author: ```suggestion
    * **If possible, add docker containers for test resources.** (You can find examples of similar containers in `assets/docker/`.) Otherwise, please document the steps to set up a suitable test environment.
```

        .. code-block:: python

            parser.addoption(
                "--no-presto",
                action="store_true",
                help="If set, suppress tests against presto",
            )

    * **In build_test_backends_list, add a variable and if clause**

        .. code-block:: python

            no_presto = metafunc.config.getoption("--no-presto")
            if not no_presto:
                presto_conn_str = "presto://presto@localhost/memory/test_ci"
                try:
                    engine = sa.create_engine(presto_conn_str)
                    conn = engine.connect()
                except (ImportError, sa.exc.SQLAlchemyError):
                    raise ImportError(
                        "presto tests are requested, but unable to connect to the presto database at "
                        f"'{presto_conn_str}'"
                    )
                test_backends += ["presto"]

    * **In the sa fixture method, add your test_backend to the list of backends**

        .. code-block:: python

            if "postgresql" not in test_backends and "sqlite" not in test_backends and "presto" not in test_backends:

>     * **For each of the test datasets, add a schema entry for the dialect.**

Reviewer 2: This was actually quite painful to do, would it be possible to have a generic schema on each test that is translated using the mapping table instead?

Author: Created #1865 to capture this idea for later.

4. Make changes to tests/test_utils.py
######################################

    * **Add a try-except clause to import dialect-specific types and map them to generic types.**

        .. code-block:: python

            try:
                import sqlalchemy.types as sqltypes
                from pyhive.sqlalchemy_presto import presto as prestotypes
                from pyhive.sqlalchemy_presto import PrestoDialect as prestodialect

                PRESTO_TYPES = {
                    "VARCHAR": sqltypes.VARCHAR,
                    "TEXT": sqltypes.VARCHAR,
                    "CHAR": sqltypes.CHAR,
                    "DOUBLE": prestotypes.DOUBLE,
                    "INTEGER": sqltypes.INTEGER,
                    "SMALLINT": sqltypes.SMALLINT,
                    "BIGINT": sqltypes.BIGINT,
                    "DATETIME": sqltypes.TIMESTAMP,
                    "TIMESTAMP": sqltypes.TIMESTAMP,
                    "DATE": sqltypes.DATE,
                    "FLOAT": prestotypes.DOUBLE,
                    "BOOLEAN": prestotypes.BOOLEAN,
                }
            except ImportError:
                PRESTO_TYPES = {}

>     * **In get_dataset, add an elif clause to connect to data, .** Note: some SQL backends require a schema. If so, this is also the right place to create a test schema, usually called ``test_ci``.

Reviewer 2: We might want to refactor schema creation to outside of get_dataset because there are enough `get_dataset` calls that it matters

Author: Added #1866

        .. code-block:: python

            elif dataset_type == "presto":
                from sqlalchemy import create_engine

>                 engine = create_engine("presto://presto@localhost/memory/test_ci", echo=True)

Reviewer 2: I've set the echo to False now (that's what we probably want when we merge the tests)

Author: ```suggestion
                engine = create_engine("presto://presto@localhost/memory/test_ci", echo=False)
```

                conn = engine.connect()

                sql_dtypes = {}
                if (
                    schemas
                    and "presto" in schemas
                    and isinstance(engine.dialect, prestodialect)
                ):
                    schema = schemas["presto"]
                    sql_dtypes = {col: PRESTO_TYPES[dtype] for (col, dtype) in schema.items()}
                    for col in schema:

>                         type_ = schema[col]

Reviewer 2: I refactored this part to do the `.lower()` here

                        if type_.lower() in ["integer", "smallint", "bigint"]:
                            df[col] = pd.to_numeric(df[col])
                        elif type_.lower() in ["float", "double"]:
                            df[col] = pd.to_numeric(df[col])
                        elif type_.lower() in ["timestamp", "datetime"]:

>                             df[col] = pd.to_datetime(df[col]).dt.strftime('%Y-%m-%d')

Reviewer 2: also added the time here

Author: ```suggestion
                            df[col] = pd.to_datetime(df[col]).dt.strftime('%Y-%m-%d %H:%M:%S')
```

   - develop a CLI integration for this dialect

Additional notes
----------------

Future work:

* Factor ``_identify_dialect`` out of ``SqlAlchemyDataset.__init__`` into its own function.
* Instead of only including pytest options to turn off tests (i.e. include everything by default and blacklist exceptions), add the ability to selectively turn on tests (i.e. exclude everything by default and whitelist exceptions). This would make it much easier to only run tests for a specific backend when developing for that backend.
* Refactor the ``temporary_notimplemented_list`` to be part of ``DataAsset``, so that it can actually control which Expectations are run.

    * This would allow developers to start development of a new backend without having to immediately implement every single Expectation.
    * Essentially, this would decouple setting up schemas, types and fixtures from implementation of all Expectations.
    * It would also pave the way to developing an automated test grid for Expectations by backend. (We should also fetch the info from ``suppress_test_for`` and ``only_for`` from the test fixtures and include it in the master lists.)

> * Refactor ``pytest_generate_tests`` so that data loading is done inside tests, not as part of test setup. This would enable a better dev loop for fixing typing errors.

Reviewer 2: :+1: 100%
